From bf6b3b3e53feb68b83e17a41ea37866a9421cf11 Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@st.com>
Date: Tue, 6 Jul 2021 18:08:32 +0200
Subject: [PATCH 044/954] feat(st-clock): add function to restore generic timer
 rate

Add a function to restore the CPU generic timer rate from STGEN content.
After wake-up from LPLV-Stop2, STGEN content is not lost, but generic timer
has been reset.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 215dc71e59775955184292e0256fed29af7aa797 feat(st-clock): add function to restore generic timer rate]
Signed-off-by: Pascal Paillet <p.paillet@st.com>
Change-Id: I6f91dbd051f76383e9ff1d6bb86225d373dbf33a
---
 drivers/st/clk/stm32mp_clkfunc.c     | 12 ++++++++++++
 include/drivers/st/stm32mp_clkfunc.h |  1 +
 2 files changed, 13 insertions(+)

diff --git a/drivers/st/clk/stm32mp_clkfunc.c b/drivers/st/clk/stm32mp_clkfunc.c
index 379547fd11..cff174f219 100644
--- a/drivers/st/clk/stm32mp_clkfunc.c
+++ b/drivers/st/clk/stm32mp_clkfunc.c
@@ -362,6 +362,18 @@ void stm32mp_stgen_config(unsigned long rate)
 	generic_delay_timer_init();
 }
 
+/*******************************************************************************
+ * This function restores CPU generic timer rate from the STGEN clock rate.
+ ******************************************************************************/
+void stm32mp_stgen_restore_rate(void)
+{
+	unsigned long rate;
+
+	rate = mmio_read_32(STGEN_BASE + CNTFID_OFF);
+
+	write_cntfrq_el0((u_register_t)rate);
+}
+
 /*******************************************************************************
  * This function returns the STGEN counter value.
  ******************************************************************************/
diff --git a/include/drivers/st/stm32mp_clkfunc.h b/include/drivers/st/stm32mp_clkfunc.h
index 61286b2244..64f935a162 100644
--- a/include/drivers/st/stm32mp_clkfunc.h
+++ b/include/drivers/st/stm32mp_clkfunc.h
@@ -29,6 +29,7 @@ int fdt_get_clock_id(int node);
 unsigned long fdt_get_uart_clock_freq(uintptr_t instance);
 
 void stm32mp_stgen_config(unsigned long rate);
+void stm32mp_stgen_restore_rate(void);
 void stm32mp_stgen_restore_counter(unsigned long long value,
 				   unsigned long long offset_in_ms);
 unsigned long long stm32mp_stgen_get_counter(void);
-- 
2.53.0

